File: backend/app/schemas/content.py
import json
from datetime import datetime

from pydantic import BaseModel, model_validator

from app.schemas.source import SourceResponse


class ContentBase(BaseModel):
    title: str
    original_title: str | None = None
    year: int | None = None
    country: list[str] | None = None
    genres: list[str] | None = None
    synopsis: str | None = None
    poster_url: str | None = None
    tmdb_id: int | None = None
    audience: str | None = None
    license_class: str | None = None


class ContentCreate(ContentBase):
    pass


class ContentResponse(ContentBase):
    id: str
    created_at: datetime
    updated_at: datetime
    sources: list[SourceResponse] = []

    @model_validator(mode="before")
    @classmethod
    def parse_json_fields(cls, values):
        if hasattr(values, "__dict__"):
            obj = values.__dict__
        else:
            obj = dict(values)
        for field in ("country", "genres"):
            val = obj.get(field)
            if isinstance(val, str):
                try:
                    obj[field] = json.loads(val)
                except (json.JSONDecodeError, TypeError):
                    obj[field] = []
        return obj

    model_config = {"from_attributes": True}


class ContentListResponse(BaseModel):
    items: list[ContentResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/content.ts
export interface Source {
  id: string;
  content_id: string;
  source_name: string;
  watch_mode: "in_app" | "external";
  external_url: string | null;
  stream_url: string | null;
  quality_hint: string | null;
  subtitle_languages: string[];
  availability_note: string | null;
  region_hint: string | null;
  is_verified: boolean;
  last_checked_at: string | null;
  created_at: string;
}

export interface Content {
  id: string;
  title: string;
  original_title: string | null;
  year: number | null;
  country: string[] | null;
  genres: string[] | null;
  synopsis: string | null;
  poster_url: string | null;
  tmdb_id: number | null;
  audience: string | null;
  license_class: string | null;
  vote_average?: number;
  vote_count?: number;
  runtime?: number;
  popularity?: number;
  created_at: string;
  updated_at: string;
  sources: Source[];
}

export interface ContentListResponse {
  items: Content[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface SearchParams {
  q?: string;
  genre?: string;
  country?: string;
  subtitle_lang?: string;
  watch_mode?: string;
  verified_only?: boolean;
  sort_by?: string;
  page?: number;
  page_size?: number;
}
